Button: add `:disabled` selector to reset hover color for disabled buttons

ramonjd opened the pull request:
Probably resolves https://github.com/WordPress/gutenberg/issues/53177


## What?
Adds a `:disabled:hover` selector to reset hover colors for disabled Buttons.

```css
	// Unset some hovers, instead of adding :not specificity.
	&:disabled:hover,
	&[aria-disabled="true"]:hover {
		color: initial;
	}
```

## Why?

The`<Button />` text color does not inherit `:hover` state colors if it has an `aria-disabled` attribute. 

But it does not cater for the `"disabled"` attribute, which many buttons use without the additional `aria-disabled` attribute.

[This rule](https://github.com/WordPress/gutenberg/blob/fc6000ac74b3fb6dd93b142e447eb5cc189fdc35/packages/components/src/button/style.scss#L30-L30)

```css
	// Unset some hovers, instead of adding :not specificity.
	&[aria-disabled="true"]:hover {
		color: initial;
	}
```

## Testing Instructions

Here are a few screen recordings

The new selector will affect all buttons that have disabled states.

For example, the preview icon in the post editor. Add a new post but don't add any content.

Also the Reset revisions button in the site editor (it's disabled once pressed).

https://github.com/WordPress/gutenberg/assets/6458278/55f33925-7712-484f-ae29-5ec4f3b1e1de

https://github.com/WordPress/gutenberg/assets/6458278/561fc477-1c78-4a6f-9d26-b910785d0cef

To test that this PR fixes https://github.com/WordPress/gutenberg/issues/53177:

Disable code editing. The quick way is to toggle the default editor settings:

```diff
diff --git a/packages/editor/src/store/defaults.js b/packages/editor/src/store/defaults.js
index 38b79ad9a8..315484289e 100644
--- a/packages/editor/src/store/defaults.js
+++ b/packages/editor/src/store/defaults.js
@@ -25,6 +25,6 @@ export const EDITOR_SETTINGS_DEFAULTS = {
 	...SETTINGS_DEFAULTS,
 
 	richEditingEnabled: true,
-	codeEditingEnabled: true,
+	codeEditingEnabled: false,
 	enableCustomFields: undefined,
 };
```

Head over to the post editor and open the settings menu at the top right of the page using the ellipsis menu.

See that the color does not change on the disabled button.

## Screenshots or screencast <!-- if applicable -->

### Before

https://github.com/WordPress/gutenberg/assets/6458278/9049f865-05e8-4f2e-ad16-17687dd0649b

### After


https://github.com/WordPress/gutenberg/assets/6458278/8a9d6788-90b3-485c-beae-394b432ecd84

ramonjd committed:
Adding aria-disabled to the button when disabled is `true`

ramonjd committed:
Update CHANGELOG.md

jameskoster reviewed (approved):
Works. Thank you :)

alexstine commented:
Accessibility impact with adding `aria-disabled`? Styles should not be based off ARIA attributes unless the said attribute is used/needed. Without testing this, I'm a little unclear on why this is needed.

ramonjd commented:
> Styles should not be based off ARIA attributes unless the said attribute is used/needed. Without testing this, I'm a little unclear on why this is needed.

Thanks for the feedback @alexstine 

The button component has a rule that resets the hover color using an `aria-disabled` selector. 
Do you think it would be more appropriate to change that rule in the button component to use the `:disabled` selector?

Actually, now I look at the button component CSS more closely, it might be more consistent. There's already a `&:focus:not(:disabled)` rule. See: https://github.com/WordPress/gutenberg/blob/fc6000ac74b3fb6dd93b142e447eb5cc189fdc35/packages/components/src/button/style.scss#L37C2-L37C24

It would have the same effect.

ramonjd committed:
Rolling back changes to menu items
Adding disabled style to button's disabled state (and not just using the aria-label selector)

ramonjd committed:
Update CHANGELOG.md

ramonjd commented:
The scope of this PR has widened slightly. Now the hover color will not show for any buttons with a `"disabled="true"` attribute.

I think it's a safe change and was probably left out initially?

I didn't change anything else to keep the scope of this PR smaller, since there are buttons that have either the `aria-disabled` OR `disabled` prop.

tellthemachines reviewed (approved):
This makes sense to me; disabled buttons shouldn't have hover styles at all.
